File: Sample/Scripts/Global/Gameplay/GameplayEffectSpec.lua
local GameplayEffectSpec = {}

GameplayEffectSpec.effect = nil
GameplayEffectSpec.eventData = nil
GameplayEffectSpec.stacks = 1
GameplayEffectSpec.sourceKey = nil

function GameplayEffectSpec:init(effect, eventData, stacks, sourceKey)
    self.effect = assert(effect, "Gameplay Effect is required")
    self.eventData = eventData
    self.stacks = stacks or 1
    self.sourceKey = sourceKey
    assert(math.type(self.stacks) == "integer" and self.stacks > 0, "Gameplay Effect stacks must be positive")
end

return class(GameplayEffectSpec)

File: Game/Engine/Source/Global/src/Gameplay/GameplayEffectSpec.cpp
#include <Gameplay/GameplayEffectSpec.hpp>
#include <Gameplay/GameplayEventData.hpp>
#include <Gameplay/GameplayEffect.hpp>

#include <cmath>
#include <stdexcept>
#include <utility>

GameplayEffectSpec::GameplayEffectSpec(
    std::shared_ptr<GameplayEffect> gameplayEffect,
    std::shared_ptr<GameplayEventData> gameplayEvent, int stackCount,
    RuntimeValue source)
    : effect(std::move(gameplayEffect)),
      eventData(std::move(gameplayEvent)),
      stacks(stackCount),
      sourceKey(std::move(source)) {
    if (effect == nullptr) {
        throw std::invalid_argument("Gameplay Effect is required");
    }
    if (stacks <= 0) {
        throw std::invalid_argument("Gameplay Effect stacks must be positive");
    }
}
